### PR TITLE
✨ Clusterctl alpha rollout status

### PR DESCRIPTION
**Pull request by hiromu-a5a**
**What this PR does / why we need it**:

In the current implementation, it's hard to track the status of rollout for a specific resource.
The basic behavior of this status command is similar to the `kubectl rollout status` which shows a message ever time rollout progress. A few differences are: i) only the machinedeployment/kubeadmcontrolplane are supported; ii) `--revision=<revision>` option is omitted since when setting this option in kubectl, it just shows an error which is not informative; iii) As a business logic, codes in this PR directly GET resources whereas kubectl uses informer, because using the informer only for this command is too much; iv) timeout is hard coded whereas kubeclt use `.spec.progressDeadlineSeconds`, because `.spec.progressDeadlineSeconds` in MachineDeployment doesn't have any effect.

In other words, this PR is the minimum implementation to satisfy most usecases of `rollout status` command.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Tracking Issue:
https://github.com/kubernetes-sigs/cluster-api/issues/3439

### CODE REVIEW

**Comment by Jont828** [comment]
@hiromu-a5a Hey, are you still working on this? It looks like it needs a rebase and I'm happy to give you a review once that's done.

**Comment by hiromu-a5a** [comment]
> @hiromu-a5a Hey, are you still working on this? It looks like it needs a rebase and I'm happy to give you a review once that's done.

Yes. thank you for reminding me. I'll do and let you know.

**Comment by vincepri** [comment]
@hiromu-a5a Any updates on this PR?

**Comment by hiromu-a5a** [comment]
@vincepri @Jont828 My apologies for the late reply. I've rebased it. Could you review it again?

File: cmd/clusterctl/client/alpha/rollout_statuswatcher.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	Status(ctx context.Context, proxy cluster.Proxy, name, namespace string) (string, bool, error)
+}
+
+// KubeadmControlPlaneStatusWatcher implements the StatusViewer interface.
+type KubeadmControlPlaneStatusWatcher struct{}
+
+// MachineDeploymentStatusWatcher implements the StatusViewer interface.
+type MachineDeploymentStatusWatcher struct{}
+
+// ObjectStatusWatcher will issue a view on the specified cluster-api resource.
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_statuswatcher.go:43]:
```suggestion
// ObjectStatusWatcher will issue a view on the specified Cluster API resource.
```

File: cmd/clusterctl/client/alpha/rollout_statuswatcher.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	case KubeadmControlPlane:
+		statusViewer = &KubeadmControlPlaneStatusWatcher{}
+	case MachineDeployment:
+		statusViewer = &MachineDeploymentStatusWatcher{}
+	default:
+		return errors.Errorf("invalid resource type %q, valid values are %v", ref.Kind, validResourceTypes)
+	}
+
+	oldStatus := ""
+	timeout := 600 * time.Second
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_statuswatcher.go:57]: Is there a shared timeout defined in a const somewhere?

File: cmd/clusterctl/client/alpha/rollout_statuswatcher.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// Status returns a status describing kubeadmcontrolplane status, and a bool value indicating if the status is considered done.
+func (s *KubeadmControlPlaneStatusWatcher) Status(ctx context.Context, proxy cluster.Proxy, name, namespace string) (string, bool, error) {
+	newKcp, err := getKubeadmControlPlane(ctx, proxy, name, namespace)
+	if err != nil || newKcp == nil {
+		return "", false, errors.Wrapf(err, "failed to get kubeadmcontrolplane/%v", name)
+	}
+	if newKcp.Spec.Replicas != nil && newKcp.Status.UpdatedReplicas < *newKcp.Spec.Replicas {
+		return fmt.Sprintf("Waiting for kubeadmcontrolplane %q rollout to finish: %d out of %d new replicas have been updated...\n", newKcp.Name, newKcp.Status.UpdatedReplicas, *newKcp.Spec.Replicas), false, nil
```
Comment (Jont828) [inline cmd/clusterctl/client/alpha/rollout_statuswatcher.go:86]: Q: Is there a way to reuse something like the ready condition in the status? That should already provide a summary of the resource as well as info on if it's an error, warning, or informational message.

**Comment by Jont828** [comment]
@hiromu-a5a Are you still working on this? If not we might want to close this PR as it's been without activity for almost 6 months.

**Comment by sbueringer** [comment]
At this point it's very very likely that we will deprecate and remove the revision management, see https://github.com/kubernetes-sigs/cluster-api/issues/10479

**Comment by hiromu-a5a** [comment]
I am really sorry for the late reply. We are shifting to GitOps and lost motivation to clusterctl.
